pkg/controller/operators/catalog/operator.go

```diff
@@ -1418,9 +1453,79 @@ func (o *Operator) ExecutePlan(plan *v1alpha1.InstallPlan) error {
 	b := newBuilder(kubeclient, dynamicClient, o.csvProvidedAPIsIndexer, o.logger)
 
+	refForStep := func(step *v1alpha1.Step) *UnpackedBundleReference {
+		log := o.logger.WithFields(logrus.Fields{"resolving": step.Resolving, "step": step.Resource.Name})
+		var ref UnpackedBundleReference
+		if err := json.Unmarshal([]byte(step.Resource.Manifest), &ref); err != nil {
+			log.Debug("step is not a reference to an unpacked bundle (this is not an error if the step is a manifest)")
+			return nil
+		}
+		log = log.WithField("ref", ref)
+		if ref.Kind != "ConfigMap" || ref.Name == "" || ref.Namespace == "" || ref.CatalogSourceName == "" || ref.CatalogSourceNamespace == "" {
+			log.Debug("step is not a reference to an unpacked bundle (this is not an error if the step is a manifest)")
+			return nil
+		}
+		return &ref
+	}
+
+	// for each step that points to a configmap instead of holding the full step manifest, find the steps and cache them
+	unpackedSteps := map[string][]v1alpha1.StepResource{}
+	loader := configmap.NewBundleLoader()
+	for _, step := range plan.Status.Plan {
+		ref := refForStep(step)
+		if ref == nil {
+			continue
+		}
+		cm, err := o.lister.CoreV1().ConfigMapLister().ConfigMaps(ref.Namespace).Get(ref.Name)
+		if err != nil {
+			return errorwrap.Wrapf(err, "error finding unpacked bundle configmap for ref %s", step.Resource.Manifest)
+		}
+		b, err := loader.Load(cm)
+		if err != nil {
+			return errorwrap.Wrapf(err, "error loading unpacked bundle configmap for ref %s", step.Resource.Manifest)
+		}
+		steps, err := resolver.NewStepResourceFromBundle(b, plan.GetNamespace(), ref.Replaces, ref.CatalogSourceName, ref.CatalogSourceNamespace)
+		if err != nil {
+			return errorwrap.Wrapf(err, "error calculating steps for ref %s", step.Resource.Manifest)
+		}
+		unpackedSteps[step.Resolving] = steps
+	}
+
+	manifestForStep := func(step *v1alpha1.Step) (string, error) {
+		manifest := step.Resource.Manifest
+		ref := refForStep(step)
+		if ref != nil {
+			log := o.logger.WithFields(logrus.Fields{"resolving": step.Resolving, "step": step.Resource.Name})
+			log.WithField("ref", ref).Debug("step is a reference to configmap")
+			if usteps, ok := unpackedSteps[step.Resolving]; ok {
+				log.Debugf("checking cache for unpacked step")
+				// need to find the real manifest from the unpacked steps
+				for _, u := range usteps {
+					if u.Name == step.Resource.Name &&
+						u.Kind == step.Resource.Kind &&
+						u.Version == step.Resource.Version &&
+						u.Group == step.Resource.Group {
+						manifest = u.Manifest
+						log.WithField("manifest", manifest).Debug("step replaced with unpacked value")
+					}
+				}
+			}
+			if manifest == step.Resource.Manifest {
+				return "", errorwrap.Wrapf(err, "couldn't find unpacked step for %v", step)
+			}
+		}
+		return manifest, nil
+	}
+
 	for i, step := range plan.Status.Plan {
+		// TODO: should lazy load only when the manifest is actually needed
+		manifest, err := manifestForStep(step)
```

Comment:
nit: WDYT about having a `type Manifest string` and using that in the code? helps improve readability, and we can hang methods off the type

Reply:
Since the type on the API is `string`, we'd have to handle manifests that are already unpacked differently from manifests that are in configmaps - instead I'll move all manifest handling into its own file, which should make it easier to keep track of what's what?